### PR TITLE
Gender NeoPronoun Transform

### PR DESCRIPTION
**Pull request by jjangsangy**
Neopronouns are a category of new (neo) pronouns that are increasingly used in place of “she,” “he,” or “they” when referring to a person. Some examples include: xe/xem/xyr, ze/hir/hirs, and ey/em/eir. Neopronouns can be used by anyone, though most often they are used by transgender, non-binary, and/or gender nonconforming people.

This transformation performs **_grammatically correct substitution_** of the gendered pronouns he/she in a given sentence with their neopronoun counterparts. The reason why a simple look-up-table approach might not work as depending on the context, the _case_ can belong to the _Nominative/Subject Pronouns,	Accusative/Object Pronouns,	Pronominal Possessive/Possessive Adjectives,	Predicative Possessive/Possessive Pronouns	or Reflexive Pronouns_ categories.

@vinayprabhu @rteehas @nick11roberts

### CODE REVIEW

**Comment by jjangsangy** [comment]
@vinayprabhu

File: transformations/gender_neopronouns/README.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# Neopronoun substitution 🏳️‍⚧️
+
+Authors:
```
Comment (sebastianGehrmann) [inline transformations/gender_neopronouns/README.md:3]: Would you mind adding your affiliation as well?

File: transformations/gender_neopronouns/README.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+- Nicolas Roberts \<nick11roberts@cs.wisc.edu\>
+
+<img alt=" The neopronoun flag from the LGBTA Wiki page" title="Neopronouns Flag by Geekycorn on DeviantArt." src="https://static.wikia.nocookie.net/lgbta/images/4/47/Neopronoun_Flag.png/revision/latest/scale-to-width-down/220?cb=20200425234516" width="200" height="200" />
+
+**1. What are neopronouns?**
+
+Answer: _Neopronouns are a category of new (neo) pronouns that are increasingly used in place of “she,” “he,” or “they” when
+referring to a person. Some examples include: xe/xem/xyr, ze/hir/hirs, and ey/em/eir. Neopronouns can be used by
+anyone, though most often they are used by transgender, non-binary, and/or gender nonconforming people._
+Source: https://intercultural.uncg.edu/wp-content/uploads/Neopronouns-Explained-UNCG-Intercultural-Engagement.pdf
```
Comment (sebastianGehrmann) [inline transformations/gender_neopronouns/README.md:16]: Can you format this as a [link](https://intercultural.uncg.edu/wp-content/uploads/Neopronouns-Explained-UNCG-Intercultural-Engagement.pdf)?

File: transformations/gender_neopronouns/README.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+In the image below, we can find the temporal variations in neopronoun usage over the past few years from  the **_Gender Census 2021: Worldwide Report_**.
```
Comment (sebastianGehrmann) [inline transformations/gender_neopronouns/README.md:18]: Can you add a sentence here with an interpretation of the image?

File: transformations/gender_neopronouns/README.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+## What tasks does it intend to benefit?
+As things stand, NLP models such as those deployed for neural translation fail to recognize the neopronouns and often treats them as proper nouns. For example, Google translate translates the sentence _Ve likes verself_ to ವೀ ತನ್ನನ್ನು ತಾನೇ ಇಷ್ಟಪಡುತ್ತಾನೆ into Kannada which reads _(Someone named) Vee likes himself_! Also, as revealed in this survey titled [Toward Gender-Inclusive Coreference Resolution](https://arxiv.org/pdf/1910.13913.pdf), _"Only 7.1% (one paper!) considers neopronouns and/or specific singular THEY."_
+This transformation seeks to render the training data used in NLP pipelines more neopronoun aware so as to help tackle the downstream risks of [trans erasure](https://allthingslinguistic.com/post/118373603278/i-am-at-the-end-of-my-first-semester-of).
+
+Benchmark Results:
+* Seq2Seq Generation task ran on IMDB dataset/Roberta model with a 1% sample size. The original accuracy was 96% and the perterbed accuracy was 95%
```
Comment (sebastianGehrmann) [inline transformations/gender_neopronouns/README.md:38]: nit: perterbed-> perturbed

Comment (sebastianGehrmann) [reply transformations/gender_neopronouns/README.md:38]: Also, I am unsure what you mean by Seq2Seq model on IMDB with Roberta. This is a classification task and a masked language model, both of which are not seq2seq.

File: transformations/gender_neopronouns/README.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+referring to a person. Some examples include: xe/xem/xyr, ze/hir/hirs, and ey/em/eir. Neopronouns can be used by
+anyone, though most often they are used by transgender, non-binary, and/or gender nonconforming people._
+Source: https://intercultural.uncg.edu/wp-content/uploads/Neopronouns-Explained-UNCG-Intercultural-Engagement.pdf
+
+In the image below, we can find the temporal variations in neopronoun usage over the past few years from  the **_Gender Census 2021: Worldwide Report_**.
+
+<img title=" The neopronoun preferences over times" src="https://gendercensus.files.wordpress.com/2021/03/gc2021-pronouns-over-time-minus-he-she-they.png" width="600" height="354" />
+
+## What type of a transformation is this?
+This transformation performs **_grammatically correct substitution_** of the gendered pronouns he/she in a given sentence with their neopronoun counterparts. The reason why a simple look-up-table approach might not work as depending on the context, the _case_ can belong to the _Nominative/Subject Pronouns,	Accusative/Object Pronouns,	Pronominal Possessive/Possessive Adjectives,	Predicative Possessive/Possessive Pronouns	or Reflexive Pronouns_ categories.
```
Comment (sebastianGehrmann) [inline transformations/gender_neopronouns/README.md:23]: #benderrule - Can you specifically state that this is for English?

File: transformations/gender_neopronouns/test.json [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      {
+        "class": "GenderNeoPronouns",
+        "args" :{
+          "swap_neopronoun": "Ey"
+       },
+        "inputs": {
+          "sentence": "I have to drive myself to school."
+        },
+        "outputs": [{
+          "sentence": "Ey have to drive emself to school."
```
Comment (sebastianGehrmann) [inline transformations/gender_neopronouns/test.json:37]: Can you add a test case with the third person singular s, like a transformation of "She walks to the mall" to "they"?

File: transformations/gender_neopronouns/transformation.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+                try:
+                    neopronoun = TAGS[self.swap_neopronoun][pron_type]
+                    neopronoun = (
+                        neopronoun.capitalize()
+                        if token.is_title
+                        else neopronoun.lower()
+                    )
+                    pieces.append(neopronoun)
+                except Exception:
```
Comment (sebastianGehrmann) [inline transformations/gender_neopronouns/transformation.py:76]: Please specify which type of Exception.

File: transformations/gender_neopronouns/transformation.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        self.nlp = spacy.load("en_core_web_sm")
+        self.swap_neopronoun = swap_neopronoun.capitalize()
+        # ensure swap_neopronoun exists in TAGS
+        assert (
+            self.swap_neopronoun in TAGS
+        ), f"swap_neopronoun can only have value of {TAGS.keys()}"
+
+        super().__init__(seed=seed, verbose=verbose, max_outputs=max_outputs)
+
+    def generate(self, sentence: str) -> List[str]:
```
Comment (sebastianGehrmann) [inline transformations/gender_neopronouns/transformation.py:53]: Please add a docstring

File: transformations/gender_neopronouns/transformation.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                if "Case" in morph and "Reflex" in morph:
+                    pron_type = "REF"
+                elif "Case" in morph:
+                    pron_type = morph["Case"].upper()
+                elif "Poss" in morph:
+                    pron_type = morph["PronType"].upper()
+                else:
+                    pron_type = "ACC"
+
+                try:
```
Comment (sebastianGehrmann) [inline transformations/gender_neopronouns/transformation.py:68]: I may have missed it but where is the third person singular s handling here?

**Review by sebastianGehrmann — CHANGES_REQUESTED** [review]
Fantastic PR, thank you for the contribution! I left a bunch of minor comments.

**Comment by james-simon** [comment]
I'm here to leave a review. Thanks for the timely task, and thanks @sebastianGehrmann for taking a close look at the code! This looks like a high-quality task that should end up in the benchmark. I have only a few high-level points to address.
* Your README mentions that, since the same standard pronoun is often used in multiple cases, this task needs grammar-aware substitution. It'd help clarify and emphasize this fact if you gave an example sentence (e.g. "Her dog nuzzled her") that illustrates that ambiguity.
* In the same vein, you should add a few tests that show that your code actually does identify the right pronoun cases despite that ambiguity.
* Your README states that "this transformation performs grammatically correct substitution of the gendered pronouns he/she in a given sentence with their neopronoun counterparts," but, judging by your tests, your code swaps out first-person and they/... pronouns, too. Which is it meant to do? Also, just out of curiosity, do neopronoun users often replace the standard first-person pronouns?

Let me know if you have any questions!

**Comment by jjangsangy** [comment]
@sebastianGehrmann I've gone and addressed the minor issues and documentation. I'm currently working on the issue with the first person singular as well as the wrong cases mentioned by @james-simon. Thanks for both of your helpful inputs!

File: transformations/gender_neopronouns/transformation.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+class GenderNeoPronouns(SentenceOperation):
+    """This transformation performs **_grammatically correct substitution_**
+    of the gendered pronouns he/she in a given sentence with their
+    neopronoun counterparts.
+    """
+
+    tasks = [TaskType.TEXT_TO_TEXT_GENERATION]
+    languages = ["en"]
+
```
Comment (asnota) [inline transformations/gender_neopronouns/transformation.py:19]: Please provide relevant keywords, using [this list](https://github.com/GEM-benchmark/NL-Augmenter/blob/main/docs/keywords.md)

File: transformations/gender_neopronouns/transformation.py (hunk 9 above)
Comment (jjangsangy) [reply transformations/gender_neopronouns/transformation.py:100]: I added the section for third person singular handling below in the `elif` clause.

**Comment by jjangsangy** [comment]
@james-simon thanks for the helpful comments.

* I've addressed the issue with the grammar-aware case of "Her dog nuzzled her". It now produces the correct substitution to be "Their dog nuzzled them".
* I also added 2 more additional test cases to cover this ambiguous case.

> Also, just out of curiosity, do neopronoun users often replace the standard first-person pronouns?
* First person pronouns should not be swapped, this was an oversight on my part. This is now reflected in the code and should pass through and not be replaced by a neopronoun.

**Comment by james-simon** [comment]
Thanks for the changes! For completeness, you should probably also add tests showing that 2nd-person and 1st-person-plural pronouns pass through untransformed. Trusting you'll do that (or explain why it doesn't matter), I'll now accept the task.

**Comment by jjangsangy** [comment]
Gotcha! I've implemented filtering of second person pronouns as well and wrote the test case showing you/your pass through.

**Review by sebastianGehrmann — APPROVED** [review]

**Comment by sebastianGehrmann** [comment]
Thanks for all the changes! Looking great to me know. If the tests pass, I am happy with it being merged. Thank you for your contribution!

**Comment by jjangsangy** [comment]
Tests have passed but it looks like we need 2 reviews to successfully complete the merge.

**Pull request merged**